go/database/mpt/proof.go

```diff
@@ -0,0 +1,116 @@
+	return &WitnessProof{}
+}
+
+// ExtractStorageProof extracts a sub-proof from the current witness proof.
+// It returns a copy that contains only proof for the storage key.
+// If the key is not found in the current proof, it returns an empty proof.
+func (p *WitnessProof) ExtractStorageProof(nodeSource NodeSource, storageRoot common.Hash, address common.Address, key common.Key) *WitnessProof {
+	return &WitnessProof{}
+}
+
+// ProveAccount verifies the witness proof against the input root hash and address.
+// If the witness proof represents the input account for the input address, it returns the account node.
+// If the proof is invalid or does not represent the account, it returns nil.
+func (p *WitnessProof) ProveAccount(nodeSource NodeSource, root common.Hash, address common.Address) *AccountNode {
+	path := AddressToNibblePath(address, nodeSource)
+	node := p.provePath(root, path)
+	if node == nil {
+		return nil
+	}
+	return node.(*AccountNode)
+}
+
+// ProveStorage verifies the witness proof against the input storage root hash, and storage key.
+// If the witness proof represents the input storage slot for the input key, it returns the value node.
+// If the proof is invalid or does not represent the storage slot, it returns nil.
+func (p *WitnessProof) ProveStorage(nodeSource NodeSource, storageRoot common.Hash, key common.Key) *ValueNode {
+	path := KeyToNibblePath(key, nodeSource)
+	node := p.provePath(storageRoot, path)
+	if node == nil {
+		return nil
+	}
+	return node.(*ValueNode)
+}
+
+func (p *WitnessProof) provePath(root common.Hash, path []Nibble) Node {
+	// TODO draft of the algorithm
+
+	//rlp, exists := (*p)[root]
+	//if !exists {
+	//	return nil
+	//}
+	//node, err := decode(rlp)
+	//if err != nil {
+	//	return nil // malformed RLP in proof
+	//}
+	//nextRoot, subPath := getNext(node, path)
+	//if nextRoot == empty {
+	//	return node
+	//}
+	//
+	//return p.provePath(nextRoot, subPath)
+
+	return nil
+}
+
+// CreateWitnessProof creates a witness proof for the input account address
+// and possibly storage slots of the same account under the input storage keys.
+func CreateWitnessProof(nodeSource NodeSource, root *NodeReference, address common.Address, keys ...common.Key) (*WitnessProof, error) {
+	return nil, nil
+}
+
+// proofExtractionVisitor is a visitor that visits MPT nodes and creates a witness proof.
+// It hashes and encodes the nodes and stores them into the proof database.
+type proofExtractionVisitor struct {
+	proof      *WitnessProof
+	nodeSource NodeSource
+	err        error
+}
+
+// Visit computes RLP and hash of the visited node and puts it to the proof.
+func (p *proofExtractionVisitor) Visit(node Node, _ NodeInfo) VisitResponse {
+	data := make([]byte, 0, 1024)
+	rlp, err := encode(node, p.nodeSource, data)
+	if err != nil {
+		p.err = err
+		return VisitResponseAbort
+	}
+	hash := common.Keccak256(rlp)
+
+	(*p.proof)[hash] = rlp
+
+	return VisitResponseContinue
+}
```

Comment:
How about leaning more into the concept of a `WitnessProof` as a `DataBase`, as described in the comments above. The user interaction could look somehow like this:
```

	proof := CreateProof( ... )

	valid := proof.IsValid()
	info, err := proof.GetAccountInfo(root, addr)
	balance, err := proof.GetBalance(root, addr)
	nonce, err := proof.GetNonce(root, addr)
	hash, err := proof.GetCodeHash(root, addr)
	value, err := proof.GetState(root, addr, key)

	empty, err := proof.AllStatesZero(root, addr, from, to)
	empty, err := proof.AllAddressesEmpty(root, from, to)
```
where
 - `IsValid` checks that the proof has all valid entries; this means, RLP encodings are valid encodings and their hashes match the key in the map
 - `GetXXX` tries to walk from the given root hash to the requested property; if the path could be completed, the requested property is returned; if nodes are missing, an error is returned (the proof does not cover this property)
 - `AllStatesZero` returns true if from the content of the proof it can be shown that the range key `from` to key `to` (inclusive) is all zero; if there is a non-zero value in this range, false is returned; it returns an error if there is not sufficient data in the proof to show either property
 - `AllAccountsEmpty` does the same for a range of accounts

A plus for this is that the proof only needs to be evaluated once before it can be used to extract information multiple times. We could also produce an error in case a user tries to access data without prior validation.

Wdyt?

Reply:
yes, it sounds good!

shall we return `err` or `bool`?  If the client asks for a missing information, it is not an error state.  Err state would be if the proof is not valid for instance.  We could say in the err where the path could not continue, but not sure user needs / should have this information. 

Perhaps we should even return `(X, bool, err)` - err for invalid proof, bool for missing proof for the input. 
or `(*X, err) `- returning nil for a missing element - but perhaps the version with `bool` is more golang style

Reply:
perhaps `CreateProof( ... )`  should also return `bool` to be discarded if the poof could not be created for an input at all.

Reply:
added two Getters at the moment, will add more next

Reply:
> perhaps CreateProof( ... ) should also return bool to be discarded if the poof could not be created for an input at all.

`CreateProof` should only fail if there is an IO error. Attempting to proof a key or account that is not the database is a valid use case. In case of a key, it would proof that the value is zero. I case of an account it would proof that the account is empty.

Reply:
right, but I wanted to ask more, if we should return `bool` also to show that the address/key was not found. If we do not return it, the user will not know that the input data was not proven unless he calls Getters to read the value back

Reply:
A few more ideas:

To make sure that the user of the API is considering the potential invalid proof or lack of information, and not just ignores the error accidentally, we could wrap the result into a `Optional` like container. Something like this:
```
type Certified[T any] struct {
	valid bool
	value T
}

func newCertified[T any](value T) Certified[T] {
	return Certified[T]{valid: true, value: value}
}

func newMissing[T any]() Certified[T] {
	return Certified[T]{}
}

func (c Certified[T]) IsValid() bool {
	return c.valid
}

func (c Certified[T]) Value() T {
	if !c.valid {
		panic("value is not valid")
	}
	return c.value
}
```
and then have functions like
```
func (p WitnessProof) GetBalance(root Hash, addr Address) Certified[Balance] { ... }
```
This would make properties only accessible after a check, like this
```
balance := proof.GetBalance()
if !balance.IsValid() {
  // handle missing balance
} else {
  // use balance
  _ = balance.Value()
}
```
The down-side of this is that the error handling pattern is more Go-style conform.

Basically there are three options:
 - use `Value, error` and trust that the user is checking the error
 - use `*Value` and trust that the user checks for `nil`
 - use `Certified[Value]` and trust that the user checks for validity
 
All not good options ... any other ideas?

Reply:
> shall we return err or bool? If the client asks for a missing information, it is not an error state. Err state would be if the proof is not valid for instance. We could say in the err where the path could not continue, but not sure user needs / should have this information.

I agree, if we go for this, it should be a `bool`, not an `error`. 

About paths that can not be continued:
 - if the path ends in an empty node, this means the value that one looks for is zero => valid proof for the zero value
 - if nodes are missing in the proof and thus no next step is possible, there is not enough information in the proof to know the value or its validity

There are three cases we need to support:
- the proof is valid and the property looking for is covered by the nodes in the proof => return the value
- the proof is valid, but the property is not covered => return something indicating `unknown`
- the proof is not valid => return something indicating an invalid proof

Reply:
right, this lead me to a return type `value, bool, err`

* `value returned, true, nil `- when the value was found
* `value undefined/empty, false, nil` - when the value could not be found
* `value undefined/empty, false, err != nil` - when the proof is not valid, basically it could not be parsed due to wrong RLP, not sure if missing nodes are also an invalid stare, or it is just the same as if an empty node is reached

I would prefer returning this tuple of three elements.  I think this is a natural way in golang. I know the pattern with `Optional` from Java, but I think it is too sophisticated for go. I feel go code is usually written simpler

Reply:
It just makes client code look like this:
```
value, valid, err := proof.GetBalance(...)
if err != nil {
   // handle issue
} else if !valid {
   // handle issue
}
// do something with value
```
Lots of things to be checked after every read ... even if validity was already checked before. So this might be skipped by doing it once and signalling a invalid proof as an error.

Reply:
> not sure if missing nodes are also an invalid stare, or it is just the same as if an empty node is reached

A node missing that is not supposed to be an empty node (identified by a unique hash) is an invalid -- since incomplete -- proof.

Reply:
> It just makes client code look like this:
> 
> ```
> value, valid, err := proof.GetBalance(...)
> if err != nil {
>    // handle issue
> } else if !valid {
>    // handle issue
> }
> // do something with value
> ```
> 
> Lots of things to be checked after every read ... even if validity was already checked before. So this might be skipped by doing it once and signalling a invalid proof as an error.

Yes, I fully agree with this. Just I think it is how code in golang is written, i.e. perhaps if we do it this way nobody will get surprised.

Reply:
If needed we could provide a facade/wrapper on top it working in the manner you have proposed above, but the low-level access I would keep the tedious way.